### ISSUE TITLE
Large ships large lag until piloted

### ISSUE THREAD

**Issue by FrozenBeard** [APDevTeam/Movecraft#56]
Using a 4702 water ship, right click the pilot sign, it takes a good 10 seconds to register that its a ship to say piloted. This will just cause players to spam the sign over and over until it says piloted in chat.

**Comment by cccm5** [comment]
Ya it seems theres an issue with detecting large crafts. I'll need to optomize it. Either way, after thinking about it theres no system to prevent people from spamming pilot methods for what would be the same craft, I'll add something to fix that.

**Comment by Miclebrick** [comment]
i suggest having a detection/piloting lock, doing the detecting async, and getting chunk snapshots in a sync task to get the blocks of the ship you're detecting

**Comment by Miclebrick** [comment]
(also the lock should be per person)

**Comment by cccm5** [comment]
This is partially fixed, there was a method that was doing some things that caused an immense slow down. I have removed it in favor of a much faster solution. I say partially because now I've moved the detection of crafts interior/exterior to the piloting phase. As a result, a 20k block ship takes about 7 seconds to pilot, 4 of them being due to the interior/exterior detection. Hopefully I can lower this in the future.

**Comment by Miclebrick** [comment]
2 seconds is a huge amount if it's on the main thread

**Comment by cccm5** [comment]
its not

**Comment by FrozenBeard** [comment]
It needs to be faster, players will just assume it's not working and spam click it. You will have to send a message saying "detecting ship" or something, and also change the sign Color to show its happening, like orange for detecting and then change to green once it's piloted.

**Comment by Miclebrick** [comment]
```
Miclebrick commented on Jun 30 
i suggest having a detection/piloting lock, doing the detecting async, and getting chunk snapshots in a sync task to get the blocks of the ship you're detecting
```

**Comment by Miclebrick** [comment]
seems he did the async part, maybe he did the lock too

**Comment by FrozenBeard** [comment]
Still, the user needs feedback in both chat and the sign.

**Comment by Miclebrick** [comment]
It would also be good to have a detecting message yes

**Comment by cccm5** [comment]
Upon further testing, I've come up with 2 improvements. Firstly, I've changed the algorithm to instead be based on the surface area of the bounds of a craft rather than the number of blocks in the craft. Secondly, I've changed to a more effective hashcode with reduced collisions. It appears the old hashcode was improperly implemented and as a result was causing the amortized constant time lookups of Sets to instead operate in a linear manner.

tl;dr: A 62k block craft is almost instantly piloted

**Comment by Miclebrick** [comment]
it's almost like I told you to fix the hashcode implementation months ago *:thunk:* lol

**Comment by cccm5** [comment]
ya that was my bad =p

**Comment by FrozenBeard** [comment]
> based on the surface area of the bounds of a craft
Does this mean it ignores the inside? what if there are blocks that aren't allowed? 

Sounds good you are making progress!

**Comment by cccm5** [comment]
the surface area thing was for detecting what the outside/inside of a craft is, not the actual detection. Detection works the same but faster.
